types/state_aware_marker.go

```diff
@@ -0,0 +1,154 @@
+//go:generate mockgen -package types -destination appender_mock_test.go . StateAppender
+package types
+
+import (
+	"slices"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/common/model"
+
+	_ "github.com/marcboeker/go-duckdb"
+)
+
+type marker interface {
+	AlertMarker
+	GroupMarker
+}
+
+// FushableMarker is a marker that you should flush before you attempt to read state data
+type FushableMarker interface {
+	marker
+	Flush() error
+}
+
+type StateAppender interface {
+	Append(fingerprint model.Fingerprint, state AlertState)
+	AppendInhibited(fingerprint model.Fingerprint, inhibitedBy []string)
+	Close() error
+	Flush() error
+}
+
+func NewStateAwareMarker(
+	r prometheus.Registerer,
+	appender StateAppender,
+) *StateAwareMarker {
+	return &StateAwareMarker{
+		marker:   NewMarker(r),
+		appender: appender,
+	}
+}
+
+type StateAwareMarker struct {
+	marker   marker
+	appender StateAppender
+}
+
+func (s *StateAwareMarker) SetActiveOrSilenced(
+	alert model.Fingerprint,
+	version int,
+	activeSilenceIDs,
+	pendingSilenceIDs []string,
+) {
+	currStatus := s.marker.Status(alert)
+
+	s.marker.SetActiveOrSilenced(alert, version, activeSilenceIDs, pendingSilenceIDs)
+
+	// There is a chance that marker marks it as active, so we need to check.
+	if _, _, _, isSilenced := s.marker.Silenced(alert); isSilenced {
+		if currStatus.State != AlertStateSuppressed {
+			s.appender.Append(alert, AlertStateSuppressed)
+		}
+
+		return
+	}
+
+	if currStatus.State == AlertStateSuppressed && len(currStatus.InhibitedBy) > 0 {
+		return
+	}
+
+	if currStatus.State == AlertStateActive {
+		return
+	}
+
+	s.appender.Append(alert, AlertStateActive)
+}
+
+func (s *StateAwareMarker) SetInhibited(alert model.Fingerprint, alertIDs ...string) {
+	currStatus := s.marker.Status(alert)
+
+	s.marker.SetInhibited(alert, alertIDs...)
+
+	if by, isInhibited := s.marker.Inhibited(alert); isInhibited {
+		if currStatus.State != AlertStateSuppressed {
+			s.appender.AppendInhibited(alert, by)
+
+			return
+		}
+
+		if !slices.Equal(currStatus.InhibitedBy, alertIDs) {
+			s.appender.AppendInhibited(alert, by)
+		}
+
+		return
+	}
+
+	if currStatus.State == AlertStateSuppressed || currStatus.State == AlertStateActive {
+		return
+	}
+
+	s.appender.Append(alert, AlertStateActive)
+}
+
+func (s *StateAwareMarker) Count(states ...AlertState) int {
+	return s.marker.Count(states...)
+}
+
+func (s *StateAwareMarker) Status(alert model.Fingerprint) AlertStatus {
+	return s.marker.Status(alert)
+}
+
+func (s *StateAwareMarker) Delete(alert model.Fingerprint) {
+	currStatus := s.marker.Status(alert)
+
+	s.marker.Delete(alert)
+
+	if currStatus.State == AlertStateUnprocessed {
```

Comment:
Why is this a shortcut?

Reply:
The current implementation of the marker interface, which i reuse, has a lot of noops or default actions.
For example, you can call this method and if it doesn't exist, it would be a noop. 

So if the last state that my marker knows of is unprocessed(aka it doesn't exists, memMarker returns unprocessed) then we should not update it to deleted. If its any other state, then it means an operation actually happened.

I don't want to duplicate the states, hence these checks so we don't write duplicate state.

---

Another approach would be is to re-implement the marker entirely and not re-use the existing one. But I concluded that that's not worth it 🤔